cpu/esp32/periph/gpio.c

```diff
@@ -18,505 +18,353 @@
-    if (rtc) {
-        /* enable/disable the pull-up resistor (FUN_WPU) */
-        REG_SET_CLR_BIT(mode == GPIO_IN_PU || mode == GPIO_OD_PU || mode == GPIO_IN_OD_PU,
-                        rtc->reg, BIT(rtc->pullup));
-        /* enable/disable the pull-down resistor (FUN_WPD) */
-        REG_SET_CLR_BIT(mode == GPIO_IN_PD, rtc->reg, BIT(rtc->pulldown));
+
+    if (esp_idf_gpio_set_intr_type(pin, type) != ESP_OK) {
+        return -1;
     }
 
-    return 0;
-}
+    if (!_isr_installed &&
+        esp_idf_gpio_install_isr_service(ESP_INTR_FLAG_LEVEL1) != ESP_OK) {
+        return -1;
+    }
+    _isr_installed = true;
 
-#if MODULE_PERIPH_GPIO_IRQ
+    if (esp_idf_gpio_isr_handler_add(pin, cb, arg) != ESP_OK) {
+        return -1;
+    }
+
+    gpio_irq_enable(pin);
 
-static gpio_isr_ctx_t gpio_isr_ctx_table [GPIO_PIN_NUMOF] = { };
-static bool gpio_int_enabled_table [GPIO_PIN_NUMOF] = { };
+    return (gpio_int_enabled_table[pin]) ? 0 : -1;
+}
 
-void IRAM gpio_int_handler (void* arg)
+void gpio_irq_enable(gpio_t pin)
 {
-    irq_isr_enter();
-    (void)arg;
+    DEBUG("%s: gpio=%d\n", __func__, pin);
+    assert(pin < GPIO_PIN_NUMOF);
 
-    for (unsigned i = 0; i < GPIO_PIN_NUMOF; i++) {
-        if (GPIO_REG_BIT_GET(status, status1, i)) {
-            GPIO_REG_BIT_SET(status_w1tc, status1_w1tc, i);
-            if (gpio_int_enabled_table[i] && GPIO.pin[i].int_type) {
-                gpio_isr_ctx_table[i].cb (gpio_isr_ctx_table[i].arg);
-            }
-        }
+    if (esp_idf_gpio_intr_enable(pin) == ESP_OK) {
+        gpio_int_enabled_table[pin] = true;
     }
-    irq_isr_exit();
 }
 
-int gpio_init_int(gpio_t pin, gpio_mode_t mode, gpio_flank_t flank,
-                  gpio_cb_t cb, void *arg)
+void gpio_irq_disable(gpio_t pin)
 {
-    if (gpio_init(pin, mode)) {
-        return -1;
+    DEBUG("%s: gpio=%d\n", __func__, pin);
+    assert(pin < GPIO_PIN_NUMOF);
+
+    if (esp_idf_gpio_intr_disable(pin) == ESP_OK) {
+        gpio_int_enabled_table[pin] = false;
     }
+}
 
-    gpio_isr_ctx_table[pin].cb  = cb;
-    gpio_isr_ctx_table[pin].arg = arg;
+#endif /* MODULE_PERIPH_GPIO_IRQ */
 
-    GPIO.pin[pin].int_type = flank;
-    if (flank != GPIO_NONE) {
-        gpio_int_enabled_table [pin] = (gpio_isr_ctx_table[pin].cb != NULL);
-        GPIO.pin[pin].int_ena = GPIO_PRO_CPU_INTR_ENA;
-        GPIO.pin[pin].int_type = flank;
-        GPIO.pin[pin].wakeup_enable = 1;
+#if IS_USED(MODULE_ESP_IDF_GPIO_HAL)
 
-        intr_matrix_set(PRO_CPU_NUM, ETS_GPIO_INTR_SOURCE, CPU_INUM_GPIO);
-        xt_set_interrupt_handler(CPU_INUM_GPIO, gpio_int_handler, NULL);
-        xt_ints_on(BIT(CPU_INUM_GPIO));
-    }
+static gpio_hal_context_t _gpio_hal_ctx = {
+    .dev = GPIO_HAL_GET_HW(GPIO_PORT_0)
+};
 
-    return 0;
+/*
+ * Since `gpio_hal_get_level` returns the current pin level and not the value
+ * last written to the output, we have to handle the state of each pin in a
+ * separate static variable.
+ */
+uint64_t _output = 0;
```

Comment:
```suggestion
static uint64_t _output = 0;
```
?

I wonder if a [bitfield](https://github.com/RIOT-OS/RIOT/blob/master/sys/include/bitfield.h) wouldn't be better suited here to avoid 64 bit arithmetic on every GPIO access.

Reply:
In that case we would have to define it as a union of `uint64_t` and `BITFIELD(64,...)` because the HAL function requires `uint64_t`.

Furthermore, I don't know whether all the operations that would be necessary to prepare the 8-byte mask for a singe bit and the calculations as well as the while loop over the 8-byte array in bitfield operations are more efficient than some shift and logical operations on two 32 bit registers.

Reply:
> The implementation of `gpio_ll` API

I have it on my ToDo list, but it makes no sense to implement it during the migration phase.

Reply:
Ah I missed that the vendor API uses `uint64_t` already

Reply:
A further question is what `gpio_read` should return if the GPIO is configured as output. The last value written, the current pin level if it is configured as input or anything if it is not configured as input. Some implementations return the value that was last written if the GPIO is configured as output.

Reply:
> A further question is what gpio_read should return if the GPIO is configured as output. 

I think this is not defined, so I wouldn't spend extra memory on it.

Reply:
> I think this is not defined, so I wouldn't spend extra memory on it.

Unfortunately, we have to save the output state for the `gpio_toggle` function anyway, since we cannot read back the level written to an output pin when using the vendor API.

I can try to change the code to use two `uint32_t` variables instead of one `uint64_t` variable for that purpose.

Reply:
Ok, I tried different solutions for this internal `_output` stage:

For single `uint64_t` variable (current implementation) we have less instructions but a call of `__ashldi3>:
```
400d3110 <gpio_set>:
400d3110:    004136         entry   a1, 32
400d3113:    27a032         movi    a3, 39
400d3116:    02b327         bgeu    a3, a2, 400d311c
400d3119:    fd62e5         call8   400d0748 <_assert_panic>
400d311c:    ff76a1         l32r    a10, 400d2ef4
400d311f:    f990b1         l32r    a11, 400d1760
400d3122:    ffe731         l32r    a3, 400d30c0
400d3125:    20c220         or      a12, a2, a2
400d3128:    06cfa5         call8   400d9e24 <__ashldi3>
400d312b:    0388           l32i.n  a8, a3, 0
400d312d:    1c0c           movi.n  a12, 1
400d312f:    20a8a0         or      a10, a8, a10
400d3132:    1388           l32i.n  a8, a3, 4
400d3134:    03a9           s32i.n  a10, a3, 0
400d3136:    20b8b0         or      a11, a8, a11
400d3139:    f6f5a1         l32r    a10, 400d0d10
400d313c:    13b9           s32i.n  a11, a3, 4
400d313e:    20b220         or      a11, a2, a2
400d3141:    06c0a5         call8   400d9d4c <gpio_ll_set_level>
400d3144:    f01d           retw.n
```
Using two `uint32_t` variables produces more instructions but doesn't require a call:
```
400d3118 <gpio_set>:
400d3118:    004136         entry   a1, 32
400d311b:    782c           movi.n  a8, 39
400d311d:    02bd           mov.n   a11, a2
400d311f:    02b827         bgeu    a8, a2, 400d3125
400d3122:    fd6265         call8   400d0748 <_assert_panic>
400d3125:    f81c           movi.n  a8, 31
400d3127:    0c3827         bltu    a8, a2, 400d3137
400d312a:    fffa81         l32r    a8, 400d3114
400d312d:    029d           mov.n   a9, a2
400d312f:    08a8           l32i.n  a10, a8, 0
400d3131:    0002c6         j       400d3140
400d3134:    000000         ill
400d3137:    fff681         l32r    a8, 400d3110
400d313a:    e0c292         addi    a9, a2, -32
400d313d:    0028a2         l32i    a10, a8, 0
400d3140:    01a0c2         movi    a12, 1
400d3143:    401900         ssl     a9
400d3146:    a19c00         sll     a9, a12
400d3149:    2099a0         or      a9, a9, a10
400d314c:    f6f1a1         l32r    a10, 400d0d10
400d314f:    006892         s32i    a9, a8, 0
400d3152:    06c0a5         call8   400d9d5c <gpio_ll_set_level>
400d3155:    f01d           retw.n
```
The `BITFIELD` implementation seems to be comparable to using the two `uint32_t` variables, but gives a clearer source code.
```
400d3110 <gpio_set>:
400d3110:    004136         entry    a1, 32
400d3113:    27a082         movi     a8, 39
400d3116:    20b220         or       a11, a2, a2
400d3119:    02b827         bgeu     a8, a2, 400d311f
400d311c:    fd62a5         call8    400d0748 <_assert_panic>
400d311f:    ffe881         l32r     a8, 400d30c0
400d3122:    419320         srli     a9, a2, 3
400d3125:    998a           add.n    a9, a9, a8
400d3127:    f87c           movi.n   a8, -1
400d3129:    308820         xor      a8, a8, a2
400d312c:    0009a2         l8ui     a10, a9, 0
400d312f:    1c0c           movi.n   a12, 1
400d3131:    248080         extui    a8, a8, 0, 3
400d3134:    401800         ssl      a8
400d3137:    a18c00         sll      a8, a12
400d313a:    2088a0         or       8, a8, a10
400d313d:    f6f4a1         l32r     a10, 400d0d10
400d3140:    004982         s8i      a8, a9, 0
400d3143:    06c165         call8    400d9d58 <gpio_ll_set_level>
400d3146:    f01d           retw.n
```